File: Trap_Water/trap_water.go
package trapwater

func trap(height []int) int {
	if len(height) == 0 {
		return 0
	}

	leftMax := make([]int, len(height))
	rightMax := make([]int, len(height))

	leftMax[0] = height[0]
	for i := 1; i < len(height); i++ {
		leftMax[i] = max(leftMax[i-1], height[i])
	}

	rightMax[len(height)-1] = height[len(height)-1]
	for i := len(height) - 2; i >= 0; i-- {
		rightMax[i] = max(rightMax[i+1], height[i])
	}

	water := 0
	for i := 0; i < len(height); i++ {
		water += min(leftMax[i], rightMax[i]) - height[i]
	}

	return water
}

File: Trap_Water/trap_water.js
/**
 * @param {number[]} height - The array of height bar
 * @returns {number} returns the unit area water can be trapped
 */
var trap = function (height) {
	if (height.length === 0) {
		return 0;
	}
	const bar = height.length;
	const leftMax = new Array(bar);
	const rightMax = new Array(bar);

	leftMax[0] = height[0];

	for (let i = 1; i < bar; i++) {
		leftMax[i] = Math.max(leftMax[i - 1], height[i]);
	}

	rightMax[bar - 1] = height[bar - 1];

	for (let i = bar - 2; i >= 0; i--) {
		rightMax[i] = Math.max(rightMax[i + 1], height[i]);
	}

	let water = 0;

	for (let i = 0; i < bar; i++) {
		water += Math.min(leftMax[i], rightMax[i]) - height[i];
	}

	return water;
};
